### PR TITLE
OBB: Fix distorted plotting

### LINKED ISSUE #9555: Yolov8-OBB Distorted Validation Labels
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train, Val

### Bug

Hello, 

I've playing around with the yolov8-obb model and I have to say that it works pretty well on the given labelled data (dota dataset for example). Congrats for releasing a model that is finally able to achieve this in a reasonable processing time!

However, I've then tried to make it work on my own dataset and the strange thing that I experiment is that the **validation** label are distorted, but **not the training** labels. Please see below example, where I intentionally put a huge bounding box around the ship and then put the same image in both sets (train and val) and here is what is shown: 

train image with label drawn on it => it is a rectangle, as expected
train_batch1

validation image with label drawn on it => this one has the exact same label as the train image, but we get a slightly "skewed" rectangle!  The effect is much more pronounced on other images I have but I can't show them for confidentiality reasons.
val_batch1_labels

My question is the following: is yolov8_obb using these distorted labels for validation (which would be very wrong) or is this just a display artifact that arises for whatever reason? 

The below reported issue also explains the problem but no solution was given.

https://discuss.roboflow.com/t/yolov8-obb-generated-labels-are-not-rectangles-and-do-not-match-the-original-oriented-rectangles/4770

### Environment

Ultralytics YOLOv8.1.24 🚀 Python-3.9.18 torch-2.2.1+cu118 CUDA:0 (NVIDIA GeForce RTX 3070 Laptop GPU, 7966MiB)
Setup complete ✅ (16 CPUs, 62.6 GB RAM, 375.4/431.5 GB disk)

### Minimal Reproducible Example

model = YOLO('yolov8n-obb.pt')  # load an official model
results = model.train(data=data_path, epochs=10, imgsz=640, device=0, batch=1, cfg = hyp_file)

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Laughing-q**
should resolve https://github.com/ultralytics/ultralytics/issues/9555

A summary of the plotting fix:
- The issue of incorrect visualized labels first was reported by #7506, and it's caused by the inconsistency of normalized width and height of `xywhr` format introduced by [xyxyxyxy2xywhr](https://github.com/ultralytics/ultralytics/blob/bc636d08b798412c7eecda974f0019b90566dc57/ultralytics/data/augment.py#L987) when operating on normlized `xyxyxyxy` coordinates to get normalized `xywhr`. 
(normalize `xyxyxyxy` first, then convert it to `xywhr` to get final normalized `xywhr`.)
  - It's been fixed in #7592 by converting `xywhr` to `xyxyxyxy` first before denormalizing in `plot_images` function, which made the visualizations looks much better but it still looks a little distorted in some samples.
- This PR fixed the distorting issue by explicitly normalizing `xywhr` right after `xyxyxyxy2xywhr` at the very first place to fix the inconsistency of normalized width and height. (convert `xyxyxyxy` to `xywhr` first then normalize it.)

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements in data augmentation and plotting accuracy for machine learning models.

### 📊 Key Changes
- 🔄 Adjusted normalization logic for keypoints and oriented bounding boxes (OBB) in data augmentation.
- 🔍 Improved scaling and positioning logic for bounding boxes during image plotting.

### 🎯 Purpose & Impact
- These changes aim to provide more accurate and consistent input data by ensuring proper normalization of keypoints and OBBs. This can help in improving model performance, especially in tasks like object detection and pose estimation.
- By refining how bounding boxes are scaled and positioned on images, the updated plotting function enhances the visual accuracy of model predictions. This is particularly useful for analysis and debugging by developers, as well as for presentations to stakeholders without technical expertise.
- Overall, these updates contribute to making the developmental and operational aspects of working with machine learning models more robust and user-friendly.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced data normalization and bounding box handling in Ultralytics library.

### 📊 Key Changes
- Conditional normalization logic for keypoints and oriented bounding boxes (OBB) has been moved and updated for clarity and efficiency.
- Normalization of keypoints and OBB now occurs if the `normalize` flag is set, ensuring width and height consistency.
- Adjusted the image plotting logic to correctly apply scaling and positioning before converting boxes to their required format, improving accuracy in visualization.

### 🎯 Purpose & Impact
- **Better Data Consistency:** By refining when and how normalization occurs, this PR ensures that all parts of the data (keypoints, bounding boxes) are consistently scaled. This leads to more accurate training and evaluation, benefiting various applications, from object detection to pose estimation. 📏
- **Enhanced Visualization:** The reordering of operations in the plotting function ensures that visualizations accurately reflect the pre-processed state of the images and annotations. This improvement helps in debugging and in presenting more precise visual results to users. 🎨
- **Overall Impact:** These changes contribute to the robustness and usability of the Ultralytics library, potentially leading to better model performances and more reliable outputs for users across different domains. 🚀

### CODE REVIEW

**Comment by Laughing-q** [comment]
@CedricLienhard @glenn-jocher visualized results:
before:
pic-240409-1236-53
after:
pic-240409-1237-04

**Comment by zyiiiiiii** [comment]
@Laughing-q Hello, after downloading your modified augment.py file and plotting.py file to local replace the previous file,  I retrained the model, but the plotting is still distorted in val, why is that?

**Comment by Laughing-q** [comment]
@zyiiiiiii The issue has been solved in my tests. Please make sure you use this branch and try the following step to install the package:
```bash
pip install git+https://github.com/ultralytics/ultralytics@fix-plot
```
If the issue still happens to you, you could send me a minimal sample(the image and its label that you got distorted visualizations) if you don't mind, so I'm able to reproduce. :)

**Comment by zyiiiiiii** [comment]
@Laughing-q 

> @zyiiiiiii The issue has been solved in my tests. Please make sure you use this branch and try the following step to install the package:
> 
> ```shell
> pip install git+https://github.com/ultralytics/ultralytics@fix-plot
> ```
> 
> If the issue still happens to you, you could send me a minimal sample(the image and its label that you got distorted visualizations) if you don't mind, so I'm able to reproduce. :)

通用场景
术语库
I have installed it according to your tips, but the problem still doesn't seem to be solved. The image is an example of val in my training process:
val_batch0_labels

**Comment by Laughing-q** [comment]
@zyiiiiiii Maybe it can't be 100% fixed among different datasets, but it should be better than before. 
Can you also try with the main branch to see if the visualizations generated from this branch looks better? 
```bash
pip install ultralytics
```

**Comment by zyiiiiiii** [comment]
@Laughing-q 
DB-AD2-084_image_06-Feb-2024_173236
[DB-AD2-081_image_06-Feb-2024_173236.txt](https://github.com/ultralytics/ultralytics/files/14917596/DB-AD2-081_image_06-Feb-2024_173236.txt)
 I post one image and its label here, it may be my operation problem, please help me to do some tests, thanks.

**Comment by Laughing-q** [comment]
@zyiiiiiii The visualized label looks good to me. :)
on main branch which looks distorted:
val_batch0_labels
on this PR which looks good:
val_batch0_labels

**Comment by zyiiiiiii** [comment]
@Laughing-q Oh, thank you very much, and I want to know if this distorted label affects the model's predictions, and if it doesn't I want to just ignore this issue.

**Comment by glenn-jocher** [comment]
@Laughing-q @zyiiiiiii should I go ahead and the PR now?

**Comment by Laughing-q** [comment]
@glenn-jocher yes please. The fix is working! :)

**Comment by Laughing-q** [comment]
> @Laughing-q Oh, thank you very much, and I want to know if this distorted label affects the model's predictions, and if it doesn't I want to just ignore this issue.

no worries, it won't affect the model predictions.:) Plus the distorted visualization should be fixed in this PR!

**Comment by zyiiiiiii** [comment]
@Laughing-q Distorted visualization in your test has indeed been solved, too impressive, just my operation problem, thank you again for answering my question, thank you!

**Comment by zyiiiiiii** [comment]
@Laughing-q  @glenn-jocher I have another question, could you give me some guidance? I am now using yolov8_obb to train the customized data set. The content of the data set is just like I posted above. After the training, I found that the model can detect the rotating target of the rectangle well, that is, the predicted box is very tight to the target. But for the square rotating target, the model detected result is not very good, the predicted box does not tight the target very well, why?

**Comment by glenn-jocher** [comment]
@Laughing-q @zyiiiiiii PR merged, thank you both for your contributions!

**Comment by Alarmod** [comment]
@glenn-jocher 
@Laughing-q 
I checked current version. Yes, results are fixed, but not input data. For examle, this input data in validate: 
image

Here result: 
image
